File: fallTime.m
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% Analysis of WF in the case of 1 pulse : calculate the PW, the raise time
% and the fall time (all in ns)
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% Last update : 27 Sept 2023 by Romain Fons
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

clear all
close all
clc

A = dir('*.csv');

addpath('C:\Users\fonsroma\Documents\FF\ConoscopeProfileProcess')
addpath('C:\Users\fonsroma\Documents\FF\ConoscopeProfileProcess\ResearchToolbox')
% addpath('C:\Users\fonsroma\Documents\Scripts Matlab\Scripts généraux')

Name = 'MZ Jean measurement 1Pulse'; % Used for the title in figure 
Cut_threshold = 1/2; % Cut FWHM (50%) for PW determination
N = 300; % For the smoothdata fonction for not taking into accompt the overshoot, very important to play with this parameter 

for i = 1:length(A)
    Filename = A(i).name;
    
    data = importdata(Filename);
    time = data(:,1)*10^9; % time in ns
    pas_time = time(2)-time(1); %ns/point
    
    intensity = data(:,2); % intensity
    Min_intensity = min(intensity);
    
    % Smooth of the data to remove the "rising" peak for the Max_intensity calculation
    Smooth_intensity = smoothdata(intensity,'gaussian',N);
    
    %     Smooth_intensity = intensity; %Line to avoid the smooth fct in case of too short pulses (sub 0.5ns) or when in the abscence of the overshoot 
    
    
    %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
    % Because of the too low resolution of the measurement, the
    % determination of one cut follow this logic : 
    % 1. find the points right before and right after the real cut (t1, t2)
    % 2. with a linear extrapolation between these two points we find the
    % exact cut position at the exact cut (T1).
    %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
    
    % Calculation of the real PW (ns)
    Max_intensity = max(Smooth_intensity);
    threshold = Max_intensity*Cut_threshold;
    
    % Pulse width
    PW_points = [find(intensity > threshold,1,'first'),find(intensity > threshold,1,'last')];
    t1 = time(PW_points(1)-1);
    t2 = time(PW_points(1));
    Y1 = intensity(PW_points(1)-1);
    Y2 = intensity(PW_points(1));
    T1_PW = (threshold*(t1-t2)-t1*Y2+t2*Y1)/(Y1-Y2);
    
    t3 = time(PW_points(2));
    t4 = time(PW_points(2)+1);
    Y3 = intensity(PW_points(2));
    Y4 = intensity(PW_points(2)+1);
    T2_PW = (threshold*(t3-t4)-t3*Y4+t4*Y3)/(Y3-Y4);
    
    PW_ns(i) = abs(T2_PW-T1_PW);
    
    PW_ns_str = num2str(PW_ns(i),5);
    
    % Raise time 20-80 : cut1=20% - cut2=80%
    cut1 = 0.2;
    cut2 = 0.8;
   
    Raise_points = [find(intensity > Max_intensity*cut1,1,'first'),find(intensity > Max_intensity*cut2,1,'first')];
    t1 = time(Raise_points(1)-1);
    t2 = time(Raise_points(1));
    Y1 = intensity(Raise_points(1)-1);
    Y2 = intensity(Raise_points(1));
    T1_Raise = (Max_intensity*cut1*(t1-t2)-t1*Y2+t2*Y1)/(Y1-Y2);
    
    t3 = time(Raise_points(2)-1);
    t4 = time(Raise_points(2));
    Y3 = intensity(Raise_points(2)-1);
    Y4 = intensity(Raise_points(2));
    T2_Raise = (Max_intensity*cut2*(t3-t4)-t3*Y4+t4*Y3)/(Y3-Y4);
    
    Raise_ns(i) = abs(T2_Raise-T1_Raise);
    
    Raise_ns_str = num2str(Raise_ns(i),5);
    
    % Fall time 20-80
    Fall_points = [find(intensity > Max_intensity*cut2,1,'last'),find(intensity > Max_intensity*cut1,1,'last')];
    t1 = time(Fall_points(1));
    t2 = time(Fall_points(1)+1);
    Y1 = intensity(Fall_points(1));
    Y2 = intensity(Fall_points(1)+1);
    T1_Fall = (Max_intensity*cut2*(t1-t2)-t1*Y2+t2*Y1)/(Y1-Y2);
    
    t3 = time(Fall_points(2));
    t4 = time(Fall_points(2)+1);
    Y3 = intensity(Fall_points(2));
    Y4 = intensity(Fall_points(2)+1);
    T2_Fall = (Max_intensity*cut1*(t3-t4)-t3*Y4+t4*Y3)/(Y3-Y4);
    
    Fall_ns(i) = abs(T2_Fall-T1_Fall);
    
    Fall_ns_str = num2str(Fall_ns(i),5);
 
    % Plot de la figure avec les points des différents cuts
    f = figure;
    plot(time, intensity);
    grid on
    grid minor
    xlim([time(1) time(end)]);
    ylim([Min_intensity-0.005 max(intensity)*1.1]);
    title(Name);
    xlabel('time (ns)');
    ylabel('Intensity (V)');
    hold on
    plot(T1_Raise,Max_intensity*cut1,'b*')
    plot(T2_Raise,Max_intensity*cut2,'b*')
    plot(T1_Fall,Max_intensity*cut2,'r*')
    plot(T2_Fall,Max_intensity*cut1,'r*')
    plot(T1_PW,threshold,'g*')
    plot(T2_PW,threshold,'g*')
    xline(T1_PW,'g')
    xline(T2_PW,'g')
    hold off
    legend(PW_ns_str)
    title(Filename,'Interpreter', 'none');
    
    
    saveas(gcf,[Filename, '_PW_',PW_ns_str,'ns','_Raise_',Raise_ns_str,'ns','_Fall_',Fall_ns_str,'ns','.png'])
    
    
end


PW_ns
Raise_ns
Fall_ns

File: fallTime.py
import os
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from scipy.signal import find_peaks
from scipy.interpolate import interp1d

# Set the path to the directory containing CSV files
data_path = '/path/to/your/data/directory'

# Get a list of CSV files in the directory
files = [f for f in os.listdir(data_path) if f.endswith('.csv')]

# Parameters
cut_threshold = 1/2
N = 300

PW_ns = []
Raise_ns = []
Fall_ns = []

for file in files:
    # Load data
    data = pd.read_csv(os.path.join(data_path, file))
    time = data.iloc[:, 0] * 1e9  # time in ns
    pas_time = time.iloc[1] - time.iloc[0]  # ns/point

    intensity = data.iloc[:, 1]
    min_intensity = np.min(intensity)

    # Smooth data
    smooth_intensity = np.convolve(intensity, np.ones(N)/N, mode='same')

    # Pulse width calculation
    max_intensity = np.max(smooth_intensity)
    threshold = max_intensity * cut_threshold

    pw_points = find_peaks(intensity > threshold)[0][:2]
    interp_func = interp1d(time.iloc[pw_points - 1], intensity.iloc[pw_points - 1], kind='linear')
    T1_PW = interp_func.inverse(threshold)

    pw_points = find_peaks(intensity > threshold)[0][1:]
    interp_func = interp1d(time.iloc[pw_points], intensity.iloc[pw_points], kind='linear')
    T2_PW = interp_func.inverse(threshold)

    pw_ns = np.abs(T2_PW - T1_PW)
    PW_ns.append(pw_ns)

    # Raise time 20-80
    cut1 = 0.2
    cut2 = 0.8

    raise_points = find_peaks(intensity > max_intensity * cut1)[0][:2]
    interp_func = interp1d(time.iloc[raise_points - 1], intensity.iloc[raise_points - 1], kind='linear')
    T1_Raise = interp_func.inverse(max_intensity * cut1)

    raise_points = find_peaks(intensity > max_intensity * cut2)[0][:2]
    interp_func = interp1d(time.iloc[raise_points - 1], intensity.iloc[raise_points - 1], kind='linear')
    T2_Raise = interp_func.inverse(max_intensity * cut2)

    raise_ns = np.abs(T2_Raise - T1_Raise)
    Raise_ns.append(raise_ns)

    # Fall time 20-80
    fall_points = find_peaks(intensity > max_intensity * cut2)[0][-2:]
    interp_func = interp1d(time.iloc[fall_points], intensity.iloc[fall_points], kind='linear')
    T1_Fall = interp_func.inverse(max_intensity * cut2)

    fall_points = find_peaks(intensity > max_intensity * cut1)[0][-2:]
    interp_func = interp1d(time.iloc[fall_points], intensity.iloc[fall_points], kind='linear')
    T2_Fall = interp_func.inverse(max_intensity * cut1)

    fall_ns = np.abs(T2_Fall - T1_Fall)
    Fall_ns.append(fall_ns)

    # Plot the figure with the points of the different cuts
    plt.figure()
    plt.plot(time, intensity)
    plt.grid(True, which='both', linestyle='--', linewidth=0.5)
    plt.xlim([time.iloc[0], time.iloc[-1]])
    plt.ylim([min_intensity - 0.005, max_intensity * 1.1])
    plt.title(file)
    plt.xlabel('time (ns)')
    plt.ylabel('Intensity (V)')
    plt.plot(T1_Raise, max_intensity * cut1, 'b*')
    plt.plot(T2_Raise, max_intensity * cut2, 'b*')
    plt.plot(T1_Fall, max_intensity * cut2, 'r*')
    plt.plot(T2_Fall, max_intensity * cut1, 'r*')
    plt.plot(T1_PW, threshold, 'g*')
    plt.plot(T2_PW, threshold, 'g*')
    plt.axvline(T1_PW, color='g')
    plt.axvline(T2_PW, color='g')
    plt.legend([f'PW: {pw_ns:.5f} ns'])
    plt.title(file)

    plt.savefig(os.path.join(data_path, f"{file}_PW_{pw_ns:.5f}ns_Raise_{raise_ns:.5f}ns_Fall_{fall_ns:.5f}ns.png"))

# Display results
print("Pulse Widths:", PW_ns)
print("Raise Times:", Raise_ns)
print("Fall Times:", Fall_ns)
